compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -779,6 +784,45 @@ fn check_link_name_xor_ordinal(
     }
 }
 
+struct LinterStateMixedExportNameAndNoMangle {
+    export_name: Option<Span>,
+    hir_id: Option<HirId>,
+    no_mangle: Option<Span>,
+}
+
+impl LinterStateMixedExportNameAndNoMangle {
+    fn new() -> Self {
+        Self { export_name: None, hir_id: None, no_mangle: None }
+    }
+
+    fn track_export_name(&mut self, span: Span, hir_id: HirId) {
+        self.export_name = Some(span);
+        self.hir_id = Some(hir_id);
+    }
+
+    fn track_no_mangle(&mut self, span: Span) {
+        self.no_mangle = Some(span.clone());
+    }
+
+    /// Emit diagnostics if the lint condition is met.
+    fn emit_diagnostics_in_case(self, tcx: TyCtxt<'_>) {
+        if let (Some(export_name), Some(no_mangle), Some(hir_id)) =
+            (self.export_name, self.no_mangle, self.hir_id)
```

Comment:
nit: in this case we want to be exhaustive, we could use a de-structuring pattern to ensure that
```suggestion
        if let Self {
            export_name: Some(export_name),
            no_mangle: Some(no_mangle),
            hir_id: Some(hir_id),
        } = self
```